### ISSUE TITLE
perf: huge shuffle in motiff finding due to using of structs

### ISSUE THREAD

**Issue by patrick-nicholson** [graphframes/graphframes#329]
For each unique vertex and edge alias in a motif, GraphFrames currently wraps the entire vertex and edge dataframes, respectively, in aliased structs.

Spark's query planner operates on the entire contents of a struct column. This means that large column sets can be unnecessarily shuffled when they are not needed by the return statement of the query. In some cases, it will also cause dataframes to be shuffled multiple times when once would suffice.

Consider replacing the use of `struct` with dataframe aliases instead (e.g., `gf.vertices.alias(myFirstVertexAlias)`). This will allow the Spark query planner to push up column selections and reuse shuffles where possible.

This is almost certainly the cause of #322.

**Comment by patrick-nicholson** [comment]
Here's a quick demonstration:

```python
from graphframes.examples import Graphs
friends = Graphs(spark).friends()

motif_query = (
    friends.find("""
        (a) -[follows]-> (b)
    """)
    .filter("follows.relationship = 'follows'")
    .selectExpr("b.name as leader", "a.name as follower")
)

optimized_query = (
    friends.vertices.alias("a")
    .join(friends.edges.alias("follows"), F.col("a.id") == F.col("follows.src"))
    .join(friends.vertices.alias("b"), F.col("follows.dst") == F.col("b.id"))
    .filter("follows.relationship = 'follows'")
    .selectExpr("b.name as leader", "a.name as follower")
)
```

Wrapping all columns in a struct means that they're all included in the `motif_query` plan:
```python
motif_query.explain()
```
```
== Physical Plan ==
*(9) Project [b#925.name AS leader#951, a#923.name AS follower#952]
+- *(9) SortMergeJoin [follows#921.dst], [b#925.id], Inner
   :- *(6) Sort [follows#921.dst ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(follows#921.dst, 600)
   :     +- *(5) SortMergeJoin [follows#921.src], [a#923.id], Inner
   :        :- *(2) Sort [follows#921.src ASC NULLS FIRST], false, 0
   :        :  +- Exchange hashpartitioning(follows#921.src, 600)
   :        :     +- *(1) Project [named_struct(src, src#823, dst, dst#824, relationship, relationship#825) AS follows#921]
   :        :        +- *(1) Filter (named_struct(src, src#823, dst, dst#824, relationship, relationship#825).relationship = follows)
   :        :           +- Scan ExistingRDD[src#823,dst#824,relationship#825]
   :        +- *(4) Sort [a#923.id ASC NULLS FIRST], false, 0
   :           +- Exchange hashpartitioning(a#923.id, 600)
   :              +- *(3) Project [named_struct(id, id#817, name, name#818, age, age#819L) AS a#923]
   :                 +- Scan ExistingRDD[id#817,name#818,age#819L]
   +- *(8) Sort [b#925.id ASC NULLS FIRST], false, 0
      +- Exchange hashpartitioning(b#925.id, 600)
         +- *(7) Project [named_struct(id, id#817, name, name#818, age, age#819L) AS b#925]
            +- Scan ExistingRDD[id#817,name#818,age#819L]
```

The manually specified join is much more optimized, using only the columns required by the join conditions and return statement. Notice that for this example we're also able to reuse the shuffle of the vertices dataframe (because there wasn't any filtering). Above the dataframe is shuffled twice in the exact same manner.
```python
optimized_query.explain()
```
```
== Physical Plan ==
*(9) Project [name#979 AS leader#1009, name#818 AS follower#1010]
+- *(9) SortMergeJoin [dst#824], [id#978], Inner
   :- *(6) Sort [dst#824 ASC NULLS FIRST], false, 0
   :  +- Exchange hashpartitioning(dst#824, 600)
   :     +- *(5) Project [name#818, dst#824]
   :        +- *(5) SortMergeJoin [id#817], [src#823], Inner
   :           :- *(2) Sort [id#817 ASC NULLS FIRST], false, 0
   :           :  +- Exchange hashpartitioning(id#817, 600)
   :           :     +- *(1) Project [id#817, name#818]
   :           :        +- *(1) Filter isnotnull(id#817)
   :           :           +- Scan ExistingRDD[id#817,name#818,age#819L]
   :           +- *(4) Sort [src#823 ASC NULLS FIRST], false, 0
   :              +- Exchange hashpartitioning(src#823, 600)
   :                 +- *(3) Project [src#823, dst#824]
   :                    +- *(3) Filter (((isnotnull(relationship#825) && (relationship#825 = follows)) && isnotnull(src#823)) && isnotnull(dst#824))
   :                       +- Scan ExistingRDD[src#823,dst#824,relationship#825]
   +- *(8) Sort [id#978 ASC NULLS FIRST], false, 0
      +- ReusedExchange [id#978, name#979], Exchange hashpartitioning(id#817, 600)
```

**Comment by machielg** [comment]
I'm also seeing huge cost for relatively small dataset. Perhaps this is solved in  Spark 3.0? If not is there work planned in this?

**Comment by sumersao** [comment]
Is there any update on this? I'm also getting high shuffles for the .triangleCount() algorithm. Is it resolved w/ spark 3.0?

**Comment by SemyonSinchenko** [comment]
@SauronShepherd could you take a look?

**Comment by SauronShepherd** [comment]
I tested the "quick demonstration" example using the [Pokec social network](https://snap.stanford.edu/data/soc-Pokec.html) dataset. Instead of the misleading `count` action, I used:

```scala
df.write.format("noop").mode("overwrite").save()
```

I verified the results multiple times and consistently observed:

- Both the Motif and SQL-style implementations perform the same shuffle, around 32 GB.

- The Motif implementation consistently finishes 15–20 seconds earlier, indicating better performance.

**Using count to check performance in Spark is misleading** because:

**1. It only triggers a basic action:** count forces Spark to scan the DataFrame but doesn’t necessarily execute all transformations efficiently or fully, especially when dealing with complex operations like joins or motif queries.
**2. It can optimize away expensive operations:** Spark’s Catalyst optimizer may skip or simplify parts of the DAG if they don’t affect the count, so you’re not measuring the actual workload.
**3. Doesn’t reflect real shuffle or I/O costs:** Actions like write actually materialize the data, triggering all shuffles, partitions, and serialization, giving a true sense of performance.

**Bottom line:** count may finish fast, but it doesn’t accurately measure the performance of complex queries or data processing pipelines.

**Comment by patrick-nicholson** [comment]
> I tested the "quick demonstration" example using the [Pokec social network](https://snap.stanford.edu/data/soc-Pokec.html) dataset. Instead of the misleading `count` action, I used:
> 
> df.write.format("noop").mode("overwrite").save()
> 
> I verified the results multiple times and consistently observed:
> 
>     * Both the Motif and SQL-style implementations perform the same shuffle, around 32 GB.
> 
>     * The Motif implementation consistently finishes 15–20 seconds earlier, indicating better performance.
> 
> 
> **Using count to check performance in Spark is misleading** because:
> 
> **1. It only triggers a basic action:** count forces Spark to scan the DataFrame but doesn’t necessarily execute all transformations efficiently or fully, especially when dealing with complex operations like joins or motif queries. **2. It can optimize away expensive operations:** Spark’s Catalyst optimizer may skip or simplify parts of the DAG if they don’t affect the count, so you’re not measuring the actual workload. **3. Doesn’t reflect real shuffle or I/O costs:** Actions like write actually materialize the data, triggering all shuffles, partitions, and serialization, giving a true sense of performance.
> 
> **Bottom line:** count may finish fast, but it doesn’t accurately measure the performance of complex queries or data processing pipelines.

Original reporter here. FWIW, I moved on from GraphFrames pretty quickly and haven't followed Spark closely in the last 4 years or so. 

I can't be completely sure of the difference between what I was >6 years ago and what you are seeing now, especially without seeing code. What you have described might be completely correct. I think Spark query planning has changed quite a bit since the version (2.xx) available when reported. It may actually be the case the Spark is now able to detect when the shuffles are the same and reuse them.

I do want to make sure we are talking about the same thing, though.

The first, simple "quick demonstration" is not sufficient to demonstrate the shuffle implications of the original issue: there is no opportunity for shuffle reuse. I would focus more of the query plans themselves. Also, the "quick demonstration" was not a count; it was an `EXCEPT`, which requires materializing the full dataset on both sides. The count was used, along with schema comparison, to execute the query and show that the results were the same.

**Comment by SemyonSinchenko** [comment]
@SauronShepherd `find` API is the last one I did not touch.

From what I understand, packing columns to structs block projection pushdown at least. For me using table aliases is the right way. I'm going to try to experiment with aliases approach in other places of GF. Also `find` does not do any checkpointing and most probably fails on anything longer than 5-7 hops. IRL I do not think that anyone should actually look for motifs longer than 5 and `find` is not a graph query engine, but would be nice to have checkpoints anyway.

**Comment by SemyonSinchenko** [comment]
For anyone interested, take a look on this: https://graphframes.io/04-user-guide/04-motif-finding.html#performance

```
spark.conf.set("spark.sql.cbo.enabled", "true")
spark.conf.set("spark.sql.cbo.joinReorder.enabled", "true")
```

**Comment by SauronShepherd** [comment]
Hmmm, interesting, but I’d treat it as something to test rather than a guaranteed win.

Motif finding becomes a set of joins, so CBO join reordering can help. But it only helps if Spark has useful stats for edges and vertices; just enabling those configs may do nothing.

Worth trying, but I’d compare plans/benchmarks before assuming it improves performance.
